File: 931.minimum-falling-path-sum/solution.java
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;
        
        int val = 0;
        
        int srcIdx = 0;
        int dstIdx = 1;
        int MAX_INT = (int)((long)Math.pow(2, 31) - 1);
        int result = MAX_INT;
        
        int[][] dp = new int[2][];
        dp[0] = new int[m];
        dp[1] = new int[m];
        
        for (int y = 0; y < n; y++) {
            for (int x = 0; x < m; x++) {
                dp[dstIdx][x] = MAX_INT;
            }
            for (int x = 0; x < m; x++) {
                val = matrix[y][x];
                
                if (x > 0) {
                    dp[dstIdx][x - 1] = Math.min(dp[dstIdx][x - 1], dp[srcIdx][x] + val);
                }
                dp[dstIdx][x] = Math.min(dp[dstIdx][x], dp[srcIdx][x] + val);
                
                if (x + 1 < m) {
                    dp[dstIdx][x + 1] = Math.min(dp[dstIdx][x + 1], dp[srcIdx][x] + val);
                }
            }
            srcIdx = (srcIdx + 1) % 2;
            dstIdx = (dstIdx + 1) % 2;
        }
        
        for (int x = 0; x < m; x++) {
            result = Math.min(result, dp[srcIdx][x]);
        }
        
        return result;
    }
}

File: 931.minimum-falling-path-sum/solution.kt
import kotlin.math.min

class Solution {
    fun minFallingPathSum(matrix: Array<IntArray>): Int {
        val m = matrix.size
        val n = matrix[0].size
        
        val MAX_INT = (2.toDouble().pow(31) - 1).toInt()
        val dp = Array(2, { Array(n, { 0 }) })
        var result = MAX_INT
        
        var srcIdx = 0
        var dstIdx = 1
        for (y in 0 .. m - 1) {
            for (x in 0 .. n - 1) {
                dp[dstIdx][x] = MAX_INT
            }
            
            for (x in 0 .. n - 1) {
                if (x > 0) {
                    dp[dstIdx][x - 1] = min(dp[dstIdx][x - 1], dp[srcIdx][x] + matrix[y][x])
                }
                
                dp[dstIdx][x] = min(dp[dstIdx][x], dp[srcIdx][x] + matrix[y][x])
                
                if (x + 1 < m) {
                    dp[dstIdx][x + 1] = min(dp[dstIdx][x + 1], dp[srcIdx][x] + matrix[y][x])
                }
            }
            srcIdx = (srcIdx + 1) % 2
            dstIdx = (dstIdx + 1) % 2
        }
        
        for (x in 0 .. n - 1) {
            result = min(result, dp[srcIdx][x])
        }
        return result
    }
}